Simple mode implementation + mode keybindings

geksilla opened the pull request:
Implemented basic **mode** functionality.
Added function to define mode with extra info.
Prefer to use inside `init-layer!` function, for example:
```clojure
;; src/proton/layers/clojure
(defmethod init-layer! :clojure
  [_ config]
  ;; define mode name
  (mode/define-mode :clojure 
     ;; initialize for selected editor grammar
     :atom-grammars ["Clojure"] 
     ;; or initialize for file extension
     :file-extensions [#"\.proton$"]
     ;; custom init function on mode activation
     :init (fn []
            (atom-env/set-grammar "Clojure")))
  ;; define mode keybindings
  ;; also able to define within define-mode function using :mode-keybindings keyword
  ;; this one 
  (mode/define-keybindings :clojure
    {:t {:category "toggles"
         :p {:action "parinfer:toggleMode" :title "Toggle Parinfer Mode"}}})
   ;; same as
   (mode/define-mode :clojure
      :mode-keybindings {:p {:category "my-category"}})
```
Mode actions available via <kbd>SPC m</kbd>.
@dvcrn can you please review? Thanks.

dvcrn commented:
Ohh very cool! This was something I wanted do look at for a long time. I'll check it out later!

Just on the first look, maybe it makes sense to put these modes in a separate multimethod like `init-mode!` or `get-mode` or something like this

geksilla commented:
@dvcrn agree about multimethod I'll try. I was thinking about something similar to emacs modes in the future, like provide ability to compose several modes,  major/minor modes definitions etc. That's why I put it to separate file. But I'm ok with multimethod :+1:

spencerlyon2 commented:
That is so awesome! I'm excited to see this. Good work!

dvcrn commented:
Tested it and works really great :) Is this already ready for merge or do you have some more changes? 

With multimethod I meant that you currently put the stuff in `init-layer!`. What I wanted to say is that it might be better in a separate multimethod. I have no problem with the separate file :)

dvcrn commented:
Also just an idea but one thing that I think would be pretty cool is if the user can bind files to languages from the `~/.proton` file. 

Something like, I want `.abc` to always get opened in clojure mode

spencerlyon2 commented:
I've used the [file-types](https://github.com/execjosh/atom-file-types) package for this in the past

dvcrn commented:
^ might need to check if we can somehow integrate with this

I only saw inside clojure/core.cljs that he is doing something similar here:
```
(mode/define-mode :clojure
+     :atom-grammars ["Clojure"]
+     :file-extensions [#"\.proton$"]
+     :init (fn []
+            (atom-env/set-grammar "Clojure")))
```

That's why the question :stuck_out_tongue:

geksilla commented:
Basic implementation almost done. Few things left.
Need to reset modes atom state on package deactivation.
Detect and enable mode for active editor when Atom session restored. For now there is observer for active panel item which triggers on tab creation and switching between tabs. I will try to finish this one today. Thanks.

geksilla commented:
Added multimethod `describe-mode`. Example usage:
```clojure
(defmethod describe-mode :lang/clojure []
  {:mode-name :clojure
   :atom-grammars ["Clojure"]
   :file-extensions [#"\.proton$"]
   :mode-keybindings
   {:t {:category "toggles"
        :p {:action "parinfer:toggleMode" :title "Toggle Parinfer Mode"}}}
   :init clojure-mode-init})
```

Added quick access key for mode specific keybindings by <kbd>,</kbd>.
@dvcrn can you please review it? Thanks

dvcrn commented:
Cool :) I will try to check it out tomorrow. Could you also edit the [Layer Readme](https://github.com/dvcrn/proton/blob/master/HOW-TO-LAYER.md) to include your changes?

geksilla commented:
@dvcrn ok, will do.

dvcrn commented on src/cljs/proton/layers/tools/git/core.cljs:
```diff
@@ -35,3 +35,4 @@
 (defmethod get-keybindings :git [] (get-keybindings :tools/git))
 (defmethod get-initial-config :git [] (get-initial-config :tools/git))
 (defmethod init-layer! :git [] (init-layer! :tools/git))
+(defmethod describe-mode :git [] {})
```
We need to add the multimethod for `:tools/git` here. Changing this code to:
```
(defmethod describe-mode :git [] (describe-mode :tools/git))
```

Will make sure that backwards compatibility to the first version exists

dvcrn commented:
Ah I just do that myself :) Merging this now
